qmail-remote.8
.TH qmail-remote 8
.SH NAME
qmail-remote \- send mail via SMTP
.SH SYNOPSIS
.B qmail-remote
.I host
.I sender
.I recip
[
.I recip ...
]
.SH DESCRIPTION
.B qmail-remote
reads a mail message from its input
and sends the message
to one or more recipients
at a remote host.

The remote host is
.BR qmail-remote 's
first argument,
.IR host .
.B qmail-remote
sends the message to
.IR host ,
or to a mail exchanger for
.I host
listed in the Domain Name System,
via the Simple Mail Transfer Protocol (SMTP).
.I host
can be either a fully-qualified domain name:

.EX
     silverton.berkeley.edu
.EE

or an IP address enclosed in brackets:

.EX
     [128.32.183.163]
.EE

The envelope recipient addresses are listed as
.I recip
arguments to
.BR qmail-remote .
The envelope sender address is listed as
.I sender\fP.

Note that
.B qmail-remote
does not take options
and does not follow the
.B getopt
standard.
.SH TRANSPARENCY
End-of-file in SMTP is encoded as dot CR LF.
A dot at the beginning of a line is encoded as dot dot.
It is impossible in SMTP to send a message that does not end with a newline.
.B qmail-remote
converts the UNIX newline convention into the SMTP newline convention
by inserting CR before each LF.

It is a violation of the SMTP protocol
to send a message that contains long lines or non-ASCII characters.
However,
.B qmail-remote
will happily send such messages.
It is the user's responsibility to avoid generating illegal messages.
.SH "RESULTS"
.B qmail-remote
prints some number of 
.I recipient reports\fP,
followed by a
.I message report\fR.
Each report is terminated by a 0 byte.
Each report begins with a single letter:
.TP 5
r
Recipient report: acceptance.
.TP 5
h
Recipient report: permanent rejection.
.TP 5
s
Recipient report: temporary rejection.
.TP 5
K
Message report: success.
.I host
has taken responsibility for delivering the message to each
acceptable recipient.
.TP 5
Z
Message report: temporary failure.
.TP 5
D
Message report: permanent failure.
.PP
After this letter comes a human-readable description of
what happened.

The recipient reports will always be printed in the same order as
.BR qmail-remote 's
.I recip
arguments.
Note that in failure cases there may be fewer
recipient reports
than
.I recip
arguments.

.B qmail-remote
always exits zero.
.SH "CONTROL FILES"
.TP 5
.I bindroutes
Artificial interface binding routes.
Each route has the form
.IR remote-address\fB:\fIlocal-address ,
without any extra spaces.
If the address of the host to which we are delivering matches the remote
address, then bind the socket to the interface matching the local
address, making it appear as if the connection had originated from the
named local address instead of the default local address.
.I bindroutes
may contain suffix wildcards.

Send this on the internal net
.EX
   10.:10.0.0.1
.EE

Rest goes on external
.B mail
address (yes, its bogus!)
.EX
   :301.32.1.1
.EE

Hey, we could
.B cheat
on stopping qmail-remote sending to this host
.EX
   555.123.123.123:this-isnt-an-ip-address
.EE

.IR bindroutes
takes precedence over
.IR outgoingip .
.TP 5
.I domainbindings
Local IP bindings based on envelope sender domain.
Each binding has the form
.IR domain\fB:\fIlocal-ip ,
without any extra spaces.
If
.I domain
matches
.IR host ,
.B qmail-remote
will bind the local side of the socket for the outgoing connection to
.IR local-ip .
.I local-ip
may be empty;
this tells
.B qmail-remote
to let the OS choose the local address for outgoing connections as usual.
.I domainbindings
may include wildcards:

.EX
   example.com:5.6.7.8
   .example.com:5.6.7.9
   heaven.af.mil:1.2.3.4
.EE

Here the connection for any message with an envelope sender address ending with
.B .example.com
(but not
.B example.com
itself)
is bound to local address
.BR 5.6.7.9 ,
while envelope senders in
.B example.com
are bound to 
.BR 5.6.7.8 ,
and senders in
.B heaven.af.mil
are bound to 
.BR 1.2.3.4 .
All other outgoing connections (including messages with the null return-path)
use the OS-chosen default local address. The
.B qmail
system does not protect you if you specify an invalid local address to
bind to.

.IR domainbindings
takes precedence over
.IR bindroutes
and
.IR outgoingip .
.TP 5
.I helohost
Current host name,
for use solely in saying hello to the remote SMTP server.
Default:
.IR me ,
if that is supplied;
otherwise
.B qmail-remote
refuses to run.  If the envelope sender address matches a domain in the
.IR domainbindings
control file, the envelope sender domain will be used instead of this value.
If the local IP binding matches the one in the
.I helohostbindings
control file, then the matched name in the
.I helohostbindings
will be used instead.
.TP 5
.I helohostbindings
Local helo name bindings based on local IP bind used by
.IR bindroutes ,
.IR outgoingip
and
.IR domainbindings .
Each binding has the form
.IR local-ip\fB:\fIhelodomain ,
without any extra spaces.  If
.IR local-ip
matches the binded IP except 0.0.0.0 (system default),
.BR qmail-remote
will set the helo name to
.IR helodomain
instead of default
.IR me .
Example:

.EX
   5.6.7.8:host1.example.com
   5.6.7.9:host2.example.com
.EE

If there is a local IP bind used by
.IR bindroutes
,
.IR outgoingip
or
.IR domainbindings
, will trigger
.IR helohostbindings .
If local IP is bound to
.BR 5.6.7.8 ,
then
.BR qmail-remote
will set the helo name to
.BR host1.example.com .
Likewise for
.BR 5.6.7.9
will set the helo name to
.BR host2.example.com .
.TP 5
.I outgoingip
IP address to be used on outgoing connections.
Default: system-defined.
The value
.IR 0.0.0.0
is equivalent to the system default.
.TP 5
.I smtproutes
Artificial SMTP routes.
Each route has the form
.IR domain\fB:\fIrelay ,
without any extra spaces.
If
.I domain
matches
.IR host ,
.B qmail-remote
will connect to
.IR relay ,
as if
.I host
had
.I relay
as its only MX.
(It will also avoid doing any CNAME lookups on
.IR recip .)
.I host
may include a colon and a port number to use instead of the
normal SMTP port, 25:

.EX
   inside.af.mil:firewall.af.mil:26
.EE

.I relay
may be empty;
this tells
.B qmail-remote
to look up MX records as usual.
.I smtproutes
may include wildcards:

.EX
   .af.mil:
   :heaven.af.mil
.EE

Here
any address ending with
.B .af.mil
(but not
.B af.mil
itself)
is routed by its MX records;
any other address is artificially routed to
.BR heaven.af.mil .

The
.B qmail
system does not protect you if you create an artificial
mail loop between machines.
However,
you are always safe using
.I smtproutes
if you do not accept mail from the network.
.TP 5
.I timeoutconnect
Number of seconds
.B qmail-remote
will wait for the remote SMTP server to accept a connection.
Default: 60.
The kernel normally imposes a 75-second upper limit.
.TP 5
.I timeoutremote
Number of seconds
.B qmail-remote
will wait for each response from the remote SMTP server.
Default: 1200.
.SH "SEE ALSO"
addresses(5),
envelopes(5),
qmail-control(5),
qmail-send(8),
qmail-smtpd(8),
qmail-tcpok(8),
qmail-tcpto(8)


qmail-remote.c
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "sig.h"
#include "stralloc.h"
#include "substdio.h"
#include "subfd.h"
#include "scan.h"
#include "case.h"
#include "error.h"
#include "auto_qmail.h"
#include "control.h"
#include "dns.h"
#include "alloc.h"
#include "quote.h"
#include "ip.h"
#include "ipalloc.h"
#include "ipme.h"
#include "gen_alloc.h"
#include "gen_allocdefs.h"
#include "str.h"
#include "now.h"
#include "exit.h"
#include "constmap.h"
#include "tcpto.h"
#include "readwrite.h"
#include "timeoutconn.h"
#include "timeoutread.h"
#include "timeoutwrite.h"

#define HUGESMTPTEXT 5000

#define PORT_SMTP 25 /* silly rabbit, /etc/services is for users */
unsigned long port = PORT_SMTP;

GEN_ALLOC_typedef(saa,stralloc,sa,len,a)
GEN_ALLOC_readyplus(saa,stralloc,sa,len,a,i,n,x,10,saa_readyplus)
static stralloc sauninit = {0};

stralloc helohost = {0};
stralloc routes = {0};
struct constmap maproutes;
stralloc host = {0};
stralloc sender = {0};
/* for outgoing ip */
stralloc outgoingip = {0};
struct ip_address outip;
/* for domainbindings
 * References:
 *   http://pyropus.ca/software/misc/qmail-1.03-domainbindings-1.2.patch
 *   http://rno-consultores.com/mail/qmail/qmail-1.03_outgoingips.patch
 */
stralloc outdomain = {0};
stralloc senderips = {0};
struct constmap mapsenderips;
struct ip_address outip2;
/* for helohostbindings */
stralloc helohostbind = {0};
stralloc helohostname = {0};
struct constmap maphelohostbind;

saa reciplist = {0};

struct ip_address partner;

void out(s) char *s; { if (substdio_puts(subfdoutsmall,s) == -1) _exit(0); }
void zero() { if (substdio_put(subfdoutsmall,"\0",1) == -1) _exit(0); }
void zerodie() { zero(); substdio_flush(subfdoutsmall); _exit(0); }
void outsafe(sa) stralloc *sa; { int i; char ch;
for (i = 0;i < sa->len;++i) {
ch = sa->s[i]; if (ch < 33) ch = '?'; if (ch > 126) ch = '?';
if (substdio_put(subfdoutsmall,&ch,1) == -1) _exit(0); } }

void temp_badip() { out("Z\
Unable to parse IP address in control/domainbindings (#4.3.0)\n"); zerodie(); }
void temp_noip() { out("Zinvalid ipaddr in control/outgoingip (#4.3.0)\n"); zerodie(); }
void temp_nomem() { out("ZOut of memory. (#4.3.0)\n"); zerodie(); }
void temp_oserr() { out("Z\
System resources temporarily unavailable. (#4.3.0)\n"); zerodie(); }
void temp_noconn() { out("Z\
Sorry, I wasn't able to establish an SMTP connection. (#4.4.1)\n"); zerodie(); }
void temp_read() { out("ZUnable to read message. (#4.3.0)\n"); zerodie(); }
void temp_dnscanon() { out("Z\
CNAME lookup failed temporarily. (#4.4.3)\n"); zerodie(); }
void temp_dns() { out("Z\
Sorry, I couldn't find any host by that name. (#4.1.2)\n"); zerodie(); }
void temp_chdir() { out("Z\
Unable to switch to home directory. (#4.3.0)\n"); zerodie(); }
void temp_control() { out("Z\
Unable to read control files. (#4.3.0)\n"); zerodie(); }
void perm_partialline() { out("D\
SMTP cannot transfer messages with partial final lines. (#5.6.2)\n"); zerodie(); }
void perm_usage() { out("D\
I (qmail-remote) was invoked improperly. (#5.3.5)\n"); zerodie(); }
void perm_dns() { out("D\
Sorry, I couldn't find any host named ");
outsafe(&host);
out(". (#5.1.2)\n"); zerodie(); }
void perm_nomx() { out("D\
Sorry, I couldn't find a mail exchanger or IP address. (#5.4.4)\n");
zerodie(); }
void perm_ambigmx() { out("D\
Sorry. Although I'm listed as a best-preference MX or A for that host,\n\
it isn't in my control/locals file, so I don't treat it as local. (#5.4.6)\n");
zerodie(); }

void outhost()
{
  char x[IPFMT];
  if (substdio_put(subfdoutsmall,x,ip_fmt(x,&partner)) == -1) _exit(0);
}

int flagcritical = 0;

void dropped() {
  out("ZConnected to ");
  outhost();
  out(" but connection died. ");
  if (flagcritical) out("Possible duplicate! ");
  out("(#4.4.2)\n");
  zerodie();
}

int timeoutconnect = 60;
int smtpfd;
int timeout = 1200;

int saferead(fd,buf,len) int fd; char *buf; int len;
{
  int r;
  r = timeoutread(timeout,smtpfd,buf,len);
  if (r <= 0) dropped();
  return r;
}
int safewrite(fd,buf,len) int fd; char *buf; int len;
{
  int r;
  r = timeoutwrite(timeout,smtpfd,buf,len);
  if (r <= 0) dropped();
  return r;
}

char inbuf[1024];
substdio ssin = SUBSTDIO_FDBUF(read,0,inbuf,sizeof inbuf);
char smtptobuf[1024];
substdio smtpto = SUBSTDIO_FDBUF(safewrite,-1,smtptobuf,sizeof smtptobuf);
char smtpfrombuf[128];
substdio smtpfrom = SUBSTDIO_FDBUF(saferead,-1,smtpfrombuf,sizeof smtpfrombuf);

stralloc smtptext = {0};

void get(ch)
char *ch;
{
  substdio_get(&smtpfrom,ch,1);
  if (*ch != '\r')
    if (smtptext.len < HUGESMTPTEXT)
     if (!stralloc_append(&smtptext,ch)) temp_nomem();
}

unsigned long smtpcode()
{
  unsigned char ch;
  unsigned long code;

  if (!stralloc_copys(&smtptext,"")) temp_nomem();

  get(&ch); code = ch - '0';
  get(&ch); code = code * 10 + (ch - '0');
  get(&ch); code = code * 10 + (ch - '0');
  for (;;) {
    get(&ch);
    if (ch != '-') break;
    while (ch != '\n') get(&ch);
    get(&ch);
    get(&ch);
    get(&ch);
  }
  while (ch != '\n') get(&ch);

  return code;
}

void outsmtptext()
{
  int i; 
  if (smtptext.s) if (smtptext.len) {
    out("Remote host said: ");
    for (i = 0;i < smtptext.len;++i)
      if (!smtptext.s[i]) smtptext.s[i] = '?';
    if (substdio_put(subfdoutsmall,smtptext.s,smtptext.len) == -1) _exit(0);
    smtptext.len = 0;
  }
}

void quit(prepend,append)
char *prepend;
char *append;
{
  substdio_putsflush(&smtpto,"QUIT\r\n");
  /* waiting for remote side is just too ridiculous */
  out(prepend);
  outhost();
  out(append);
  out(".\n");
  outsmtptext();
  zerodie();
}

void blast()
{
  int r;
  int i;
  int o;
  char ch;
  char in[4096];
  char out[4096*2+1];
  int sol;

  for (sol = 1;;) {
    r = substdio_get(&ssin,in,sizeof in);
    if (r == 0) break;
    if (r == -1) temp_read();
    for (i = o = 0; i < r; ) {
      if (sol && in[i] == '.') {
	out[o++] = '.';
	out[o++] = in[i++];
      }
      sol = 0;
      while (i < r) {
	if (in[i] == '\n') {
	  sol = 1;
	  ++i;
	  out[o++] = '\r';
	  out[o++] = '\n';
	  break;
	}
	out[o++] = in[i++];
      }
    }
    substdio_put(&smtpto,out,o);
  }
 
  if (!sol) perm_partialline();
  flagcritical = 1;
  substdio_put(&smtpto,".\r\n",3);
  substdio_flush(&smtpto);
}

stralloc recip = {0};

void smtp()
{
  unsigned long code;
  int flagbother;
  int i;
 
  code = smtpcode();
  if (code >= 500 && code < 600) return;
  if (code >= 400 && code < 500) return; /* try next MX, see RFC-2821 */
  if (code != 220) quit("ZConnected to "," but greeting failed");
 
  substdio_puts(&smtpto,"HELO ");
  substdio_put(&smtpto,helohost.s,helohost.len);
  substdio_puts(&smtpto,"\r\n");
  substdio_flush(&smtpto);
  if (smtpcode() != 250) quit("ZConnected to "," but my name was rejected");
 
  substdio_puts(&smtpto,"MAIL FROM:<");
  substdio_put(&smtpto,sender.s,sender.len);
  substdio_puts(&smtpto,">\r\n");
  substdio_flush(&smtpto);
  code = smtpcode();
  if (code >= 500) quit("DConnected to "," but sender was rejected");
  if (code >= 400) quit("ZConnected to "," but sender was rejected");
 
  flagbother = 0;
  for (i = 0;i < reciplist.len;++i) {
    substdio_puts(&smtpto,"RCPT TO:<");
    substdio_put(&smtpto,reciplist.sa[i].s,reciplist.sa[i].len);
    substdio_puts(&smtpto,">\r\n");
    substdio_flush(&smtpto);
    code = smtpcode();
    if (code >= 500) {
      out("h"); outhost(); out(" does not like recipient.\n");
      outsmtptext(); zero();
    }
    else if (code >= 400) {
      out("s"); outhost(); out(" does not like recipient.\n");
      outsmtptext(); zero();
    }
    else {
      out("r"); zero();
      flagbother = 1;
    }
  }
  if (!flagbother) quit("DGiving up on ","");
 
  substdio_putsflush(&smtpto,"DATA\r\n");
  code = smtpcode();
  if (code >= 500) quit("D"," failed on DATA command");
  if (code >= 400) quit("Z"," failed on DATA command");
 
  blast();
  code = smtpcode();
  flagcritical = 0;
  if (code >= 500) quit("D"," failed after I sent the message");
  if (code >= 400) quit("Z"," failed after I sent the message");
  quit("K"," accepted message");
}

stralloc canonhost = {0};
stralloc canonbox = {0};

void addrmangle(saout,s,flagalias,flagcname)
stralloc *saout; /* host has to be canonical, box has to be quoted */
char *s;
int *flagalias;
int flagcname;
{
  int j;
 
  *flagalias = flagcname;
 
  j = str_rchr(s,'@');
  if (!s[j]) {
    if (!stralloc_copys(saout,s)) temp_nomem();
    return;
  }
  if (!stralloc_copys(&canonbox,s)) temp_nomem();
  canonbox.len = j;
  if (!quote(saout,&canonbox)) temp_nomem();
  if (!stralloc_cats(saout,"@")) temp_nomem();
 
  if (!stralloc_copys(&canonhost,s + j + 1)) temp_nomem();
  if (flagcname)
    switch(dns_cname(&canonhost)) {
      case 0: *flagalias = 0; break;
      case DNS_MEM: temp_nomem();
      case DNS_SOFT: temp_dnscanon();
      case DNS_HARD: ; /* alias loop, not our problem */
    }

  if (!stralloc_cat(saout,&canonhost)) temp_nomem();
}

void getcontrols()
{
  int r;
  if (control_init() == -1) temp_control();
  if (control_readint(&timeout,"control/timeoutremote") == -1) temp_control();
  if (control_readint(&timeoutconnect,"control/timeoutconnect") == -1)
    temp_control();
  if (control_rldef(&helohost,"control/helohost",1,(char *) 0) != 1)
    temp_control();
  switch(control_readfile(&routes,"control/smtproutes",0)) {
    case -1:
      temp_control();
    case 0:
      if (!constmap_init(&maproutes,"",0,1)) temp_nomem(); break;
    case 1:
      if (!constmap_init(&maproutes,routes.s,routes.len,1)) temp_nomem(); break;
  }
  r = control_readline(&outgoingip,"control/outgoingip");
  if (-1 == r) { if (errno == error_nomem) temp_nomem(); temp_control(); }
  if (0 == r && !stralloc_copys(&outgoingip, "0.0.0.0")) temp_nomem();
  if (str_equal(outgoingip.s, "0.0.0.0")) {
    outip.d[0]=outip.d[1]=outip.d[2]=outip.d[3]=(unsigned long) 0;
  }
  else if (!ip_scan(outgoingip.s, &outip)) temp_noip();
  if (outip.d[0] || outip.d[1] || outip.d[2] || outip.d[3]) {
    if (!ipme_is(&outip)) temp_noip();
  }
  switch(control_readfile(&senderips,"control/domainbindings",0)) {
    case -1:
      temp_control();
    case 0:
      if (!constmap_init(&mapsenderips,"",0,1)) temp_nomem(); break;
    case 1:
      if (!constmap_init(&mapsenderips,senderips.s,senderips.len,1)) temp_nomem(); break;
  }
  switch(control_readfile(&helohostbind,"control/helohostbindings",0)) {
    case -1:
      temp_control();
    case 0:
      if (!constmap_init(&maphelohostbind,"",0,1)) temp_nomem(); break;
    case 1:
      if (!constmap_init(&maphelohostbind,helohostbind.s,helohostbind.len,1)) temp_nomem(); break;
  }
}

void main(argc,argv)
int argc;
char **argv;
{
  static ipalloc ip = {0};
  int i;
  unsigned long random;
  char **recips;
  unsigned long prefme;
  int flagallaliases;
  int flagalias;
  char *relayhost;
  char *senderdomainip;
  char *helodomain;
  char outipstr[IPFMT];
  struct ip_address outiplocal;
  int iplocallen = 0;

  sig_pipeignore();
  if (argc < 4) perm_usage();
  if (chdir(auto_qmail) == -1) temp_chdir();
  getcontrols();
 
 
  if (!stralloc_copys(&host,argv[1])) temp_nomem();
 
  relayhost = 0;
  for (i = 0;i <= host.len;++i)
    if ((i == 0) || (i == host.len) || (host.s[i] == '.'))
      if (relayhost = constmap(&maproutes,host.s + i,host.len - i))
        break;
  if (relayhost && !*relayhost) relayhost = 0;
 
  if (relayhost) {
    i = str_chr(relayhost,':');
    if (relayhost[i]) {
      scan_ulong(relayhost + i + 1,&port);
      relayhost[i] = 0;
    }
    if (!stralloc_copys(&host,relayhost)) temp_nomem();
  }


  addrmangle(&sender,argv[2],&flagalias,0);

  /* for domainbindings
   * 'canonhost' now should contain the canonical name of the sender's host.
   * Most of the time this is a local domain name, but sometimes the domain
   * name of a forwarded email.
   */
  if(!stralloc_copy(&outdomain,&canonhost)) temp_nomem();
  senderdomainip = 0;
  for (i = 0;i <= outdomain.len;++i) {
    if ((i == 0) || (i == outdomain.len) || (outdomain.s[i] == '.')) {
      if (senderdomainip = constmap(&mapsenderips,outdomain.s + i,outdomain.len - i)) {
	break;
      }
    }
  }
  if (senderdomainip && !*senderdomainip) senderdomainip = 0;

  if (senderdomainip) {
    if (!ip_scan(senderdomainip,&outip2)) temp_badip();
    if (!stralloc_copy(&helohost,&outdomain)) temp_nomem(); /* could be in control file */
  }
  else {
    outip2.d[0] = outip2.d[1] = outip2.d[2] = outip2.d[3] = (unsigned long)0;
  }
  if (outip2.d[0] || outip2.d[1] || outip2.d[2] || outip2.d[3]) {
    if (!ipme_is(&outip2)) temp_badip();
  }

  if (!saa_readyplus(&reciplist,0)) temp_nomem();
  if (ipme_init() != 1) temp_oserr();
 
  flagallaliases = 1;
  recips = argv + 3;
  while (*recips) {
    if (!saa_readyplus(&reciplist,1)) temp_nomem();
    reciplist.sa[reciplist.len] = sauninit;
    addrmangle(reciplist.sa + reciplist.len,*recips,&flagalias,!relayhost);
    if (!flagalias) flagallaliases = 0;
    ++reciplist.len;
    ++recips;
  }

 
  random = now() + (getpid() << 16);
  switch (relayhost ? dns_ip(&ip,&host) : dns_mxip(&ip,&host,random)) {
    case DNS_MEM: temp_nomem();
    case DNS_SOFT: temp_dns();
    case DNS_HARD: perm_dns();
    case 1:
      if (ip.len <= 0) temp_dns();
  }
 
  if (ip.len <= 0) perm_nomx();
 
  prefme = 100000;
  for (i = 0;i < ip.len;++i)
    if (ipme_is(&ip.ix[i].ip))
      if (ip.ix[i].pref < prefme)
        prefme = ip.ix[i].pref;
 
  if (relayhost) prefme = 300000;
  if (flagallaliases) prefme = 500000;
 
  for (i = 0;i < ip.len;++i)
    if (ip.ix[i].pref < prefme)
      break;
 
  if (i >= ip.len)
    perm_ambigmx();
 
  for (i = 0;i < ip.len;++i) if (ip.ix[i].pref < prefme) {
    if (tcpto(&ip.ix[i].ip)) continue;
 
    smtpfd = socket(AF_INET,SOCK_STREAM,0);
    if (smtpfd == -1) temp_oserr();

    /* for domainbindings */
    if (outip2.d[0] || outip2.d[1] || outip2.d[2] || outip2.d[3]) {
      bind_by_changeoutgoingip(smtpfd, &outip2, 1);
    }

    /* for bindroutes */
    bind_by_bindroutes(smtpfd, &ip.ix[i].ip, 0);

    /* for outgoingip
     * we just need to make sure outip are not 0.0.0.0
     * which is the system defaults so no point to change it
     */
    if (outip.d[0] || outip.d[1] || outip.d[2] || outip.d[3]) {
      bind_by_changeoutgoingip(smtpfd, &outip, 0);
    }

    /* for helohostbindings */
    if (get_bind_iplocal(&outiplocal)) {
      iplocallen = ip_fmt(outipstr,&outiplocal);
      if (iplocallen > 0) {
        outipstr[iplocallen] = 0;
        if (!stralloc_copyb(&outdomain, outipstr, iplocallen)) temp_nomem();
        stralloc_0(&outdomain);
        outdomain.len--;
        helodomain = 0;
        helodomain = constmap(&maphelohostbind,outdomain.s,outdomain.len);
        if (helodomain && !*helodomain) helodomain = 0; /* no match */
        if (helodomain) { /* match */
          /* copy the helodomain into helohostname */
          if (!stralloc_copys(&helohostname, helodomain)) temp_nomem();
          if (!str_equal(helohost.s, helodomain)) {
            if (!stralloc_copy(&helohost, &helohostname)) temp_nomem();
          }
        }
      }
    }

    if (timeoutconn(smtpfd,&ip.ix[i].ip,(unsigned int) port,timeoutconnect) == 0) {
      tcpto_err(&ip.ix[i].ip,0);
      partner = ip.ix[i].ip;
      smtp(); /* only returns when the next MX is to be tried */
    }
    tcpto_err(&ip.ix[i].ip,errno == error_timeout);
    close(smtpfd);
  }
  
  temp_noconn();
}
